File: ui/ui.button.go
package ui

import "strings"

type button struct {
	size     string
	color    string
	onclick  string
	class    string
	as       string
	target   Attr
	visible  bool
	disabled bool
	attr     []Attr
}

func Button(attr ...Attr) *button {
	return &button{
		as:      "div",
		size:    MD,
		visible: true,
		target:  Target(),
		attr:    attr,
	}
}

func (b *button) Submit() *button {
	b.as = "submit"
	return b
}

func (b *button) Reset() *button {
	b.as = "reset"
	return b
}

func (b *button) If(value bool) *button {
	b.visible = value
	return b
}

func (b *button) Disabled(value bool) *button {
	b.disabled = value
	return b
}

func (b *button) Class(value ...string) *button {
	b.class = strings.Join(value, " ")
	return b
}

func (b *button) Color(value string) *button {
	b.color = value
	return b
}

func (b *button) Size(value string) *button {
	b.size = value
	return b
}

func (b *button) Click(onclick string) *button {
	// if action.Target.Id == "" {
	// 	action.Target = b.target
	// }

	b.onclick = onclick

	return b
}

func (b *button) Href(value string) *button {
	b.as = "a"
	b.attr = append(b.attr, Attr{Href: value})
	return b
}

func (b *button) Render(text string) string {
	if !b.visible {
		return ""
	}

	class := Classes(BTN, b.size, b.color, b.class, If(b.disabled, func() string { return DISABLED + " opacity-25" }))

	if b.as == "a" {
		return A(
			class,
			append(b.attr, Attr{
				ID: b.target.ID,
			})...,
		)(text)
	}

	if b.as == "div" {
		return Div(
			class,
			append(b.attr, Attr{
				ID:      b.target.ID,
				OnClick: b.onclick,
			})...,
		)(text)
	}

	return open("button")(
		class,
		append(b.attr,
			Attr{
				ID:      b.target.ID,
				Type:    b.as,
				OnClick: b.onclick,
			})...,
	)(text)
}

File: ts/ui.button.ts
import { Attr } from './ui.types';
import { BTN, Classes, DISABLED, MD } from './ui.util';
import { A, Div } from './ui.elements';

export class Button {
  private size = MD;
  private color = '';
  private onclick = '';
  private css = '';
  private as: 'div' | 'button' | 'a' = 'div';
  private target: Attr = {};
  private visible = true;
  private disabled = false;
  private extra: Attr[] = [];

  constructor(...attr: Attr[]) {
    this.extra = attr;
  }

  Submit() { this.as = 'button'; this.extra.push({ type: 'submit' }); return this; }
  Reset() { this.as = 'button'; this.extra.push({ type: 'reset' }); return this; }
  If(v: boolean) { this.visible = v; return this; }
  Disabled(v: boolean) { this.disabled = v; return this; }
  Class(...v: string[]) { this.css = v.join(' '); return this; }
  Color(v: string) { this.color = v; return this; }
  Size(v: string) { this.size = v; return this; }
  Click(code: string) { this.onclick = code; return this; }
  Href(v: string) { this.as = 'a'; this.extra.push({ href: v }); return this; }

  Render(text: string): string {
    if (!this.visible) return '';
    const cls = Classes(BTN, this.size, this.color, this.css, this.disabled && DISABLED + ' opacity-25');

    if (this.as === 'a') {
      return A(cls, ...this.extra, { id: this.target.id })(text);
    }

    if (this.as === 'div') {
      return Div(cls, ...this.extra, { id: this.target.id, onclick: this.onclick })(text);
    }

    // button element
    return `<button ${[...this.extra, { id: this.target.id, onclick: this.onclick, class: cls }]
      .map(a => Object.entries(a).map(([k, v]) => v !== undefined && v !== false ? `${k}=\"${v === true ? k : v}\"` : '').filter(Boolean).join(' ')).join(' ')}>${text}</button>`;
  }
}

export function ButtonEl(...attr: Attr[]) { return new Button(...attr); }

